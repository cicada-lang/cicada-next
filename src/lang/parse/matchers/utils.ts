import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
