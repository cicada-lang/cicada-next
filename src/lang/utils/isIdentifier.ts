const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
