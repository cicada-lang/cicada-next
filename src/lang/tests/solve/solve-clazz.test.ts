import { expect, test } from "vitest"
import { runCode } from "../utils"

test("solve Clazz", async () => {
  const output = await runCode(`

solve (A: Type, B: Type) {
  unify class { a: A, b: B } = class { a: String, b: String }
}

`)

  expect(output).toMatchInlineSnapshot('"{ A: String, B: String }"')
})

test("solve Clazz -- subclazz", async () => {
  const output = await runCode(`

solve (A: Type) {
  unify class { a: A } = class { a: String, b: String }
}

`)

  expect(output).toMatchInlineSnapshot('"{ A: String }"')
})

test("solve Clazz -- nested", async () => {
  const output = await runCode(`

solve (A: Type, B: Type) {
  unify class { c: class { a: A, b: B } } = class { c: class { a: String, b: String } }
}

`)

  expect(output).toMatchInlineSnapshot('"{ A: String, B: String }"')
})

test("solve Clazz -- occur twice", async () => {
  const output = await runCode(`

solve (A: Type, B: Type) {
  unify
    class { c: class { a: A, b: B }, b: String } =
    class { c: class { a: String, b: String }, b: B }
}

`)

  expect(output).toMatchInlineSnapshot('"{ A: String, B: String }"')
})
