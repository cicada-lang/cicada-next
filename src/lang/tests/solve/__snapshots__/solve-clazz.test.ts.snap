// Vitest Snapshot v1, https://bun:test.dev/guide/snapshot.html

exports[`solve Clazz -- nested 1`] = `
"{
  A: String,
  B: String
}"
`;

exports[`solve Clazz -- occur twice 1`] = `
"{
  A: String,
  B: String
}"
`;

exports[`solve Clazz -- subclazz 1`] = `
"{
  A: String
}"
`;

exports[`solve Clazz 1`] = `
"{
  A: String,
  B: String
}"
`;

exports[`solve ClazzCons -- deepWalk 1`] = `
"{
  A: String,
  B: String,
  C: class {
    a: String
    b: String
  }
}"
`;

exports[`solve ClazzFulfilled -- deepWalk 1`] = `
"{
  A: String,
  B: String,
  C: class {
    a: String = "a"
    b: String = "b"
  },
  a: "a",
  b: "b"
}"
`;
